Objective Function Part II - penalize disregarded utility of touched orders

bh2smith opened the pull request:
This PR is intended to provide an objective function to auction settlement which rewards total traders utility while penalizing solutions whose touched orders have disregarded utility.

In brief, the objective function is `U - dU` where `U`, the total utility, is the sum over all touched orders of their individual trade utilities `u_i`; 
```
 u_i = ((execBuyAmount * sellAmount - execSellAmount * buyAmount) * priceBuyToken) // sellAmount
```
and `dU`, the total disregarded utility, is the sum over all touched orders of their individual disregarded utilities `dU_i`; 
```
dU_i = (((sellAmount - execSellAmount) * priceBuyToken) * buyAmount) // sellAmount
```

The gas reporter gives the following estimates (for a basic trade of two orders).

```
Two Orders:
|···························································································│
|  Optimizer enabled: true  ·         Runs: 200  ·        Block limit: 6721975 gas          │
|···························································································│
|  Methods            ·               20 gwei/gas               ·       183.32 usd/eth      │
|·····················|·················|·········|········|·········|·········|············│
|                     ·  Method         ·  Min    ·  Max   ·  Avg    · # calls ·  usd (avg) │
|  With New Objective ·  submitSolution ·  410182 · 865437 ·  610996 ·  35     · 2.24       │
|  Without Objective  ·  submitSolution ·  406816 · 858665 ·  615099 ·  35     · 2.26       │
|···························································································│
```


TODO - run this same gas estimate on a larger collection of orders!

bh2smith committed:
First pass at evaluating utility and disregarded utility on solution submission.

bh2smith committed:
Compiling version of first pass

bh2smith committed:
Updated Unit tests to new objective function.

bh2smith committed:
remove Debug Events.

bh2smith committed:
evaluateUtility can be restricted to view

bh2smith committed:
typos, compile warnings and comment spacing

bh2smith committed:
include brackets in describe blocks because, for example, there are 46 occurences of "addToken" but only one of "addToken()"

bh2smith committed:
Merge branch 'cosmetics' into disregarded_utility

# Conflicts:
#	contracts/StablecoinConverter.sol
#	test/stablecoin_converter.js

fleupold commented on contracts/StablecoinConverter.sol:
```diff
@@ -209,20 +215,46 @@ contract StablecoinConverter is EpochTokenLocker {
+    function evaluateUtility(uint128 execBuy, uint128 execSell, Order memory order) internal view returns(int) {
+        // Utility = ((execBuyAmt * order.sellAmt - execSellAmt * order.buyAmt) * price.buyToken) / order.sellAmt
+        (uint128 buyAmount, uint128 sellAmount) = getBuySellAmounts(order);
+        return ((execBuy * sellAmount - execSell * buyAmount) * currentPrices[order.buyToken]) / sellAmount;
```
We need to watch out for overflow here. Should probably convert amounts to u256 and try to keep it to two multiplications (divide before multiplying with current prices) to make sure we don't overflow.

bh2smith replied:
Yes, I had this on my mind but wanted to "get something out there" for us to look at and evaluate beforehand. Not that I have already experimented with reordering the last division with the last multiplication like this: 

```py
((execBuy * sellAmt - execSell * buyAmt) / sellAmt) * price[buyToken]
```
It returns 0 in undesired places (such as our basic trade unit test, for example). The reason for this is that the magnitude of the difference obtained in the first term is usually not very large. So dividing immediately will often result in 0. My  best approach for recovering from this is to **distribute the arithmetic inside the brackets and take the difference at the very end**. Ya, that will probably work! So, new formula will be...

```py
execBuy * price[buyToken] - ((execSell * buyAmt) / sellAmt) * price[buyToken]
```

or possibly, one step simplified, 

```py
(execBuy - (execSell * buyAmt) / sellAmt) * price[buyToken]
```

bh2smith replied:
Ya, so that is nice, but the next equation (for disregarded utility) is going to be a bit more difficult to rearrange in a meaningful way without losing information...

```
(((sellAmt - execSell) * price[buyToken]) * buyAmt) / sellAmt;
```

Perhaps its best if I evaluate `maxUtility` with the utility formula above and simply take the difference 

```
dU = maxU - U
```

nlordell replied:
Don't know if this helps but could we try using [fixed point numbers](https://solidity.readthedocs.io/en/latest/types.html#fixed-point-numbers)? It looks like its not fully supported yet but maybe enough for us to use? We are operating on 128bit integers we are wasting 128 bits of precision (since, if I understand correctly, all math ops are done on 256 bit integers anyway). We could use fixed196x64 which gives us 64 bits of decimal precision and 64 extra bits to handle some overflow in the multiplications and additions. It may provide more accurate results.

bh2smith replied:
This is a cool idea, although I found the following warning at the top of the description:

```
Fixed point numbers are not fully supported by Solidity yet. 
They can be declared, but cannot be assigned to or from.
```
I'll keep this open for now and experiment in a separate sandbox!

fleupold commented on contracts/StablecoinConverter.sol:
```diff
@@ -209,20 +215,46 @@ contract StablecoinConverter is EpochTokenLocker {
+        return (((sellAmount - execSell) * currentPrices[order.buyToken]) * buyAmount) / sellAmount;
+    }
+
+    function getBuySellAmounts(Order memory order) internal pure returns(uint128, uint128) {
```
nit BuyAndSellAmounts?

fleupold commented on contracts/StablecoinConverter.sol:
```diff
@@ -209,20 +215,46 @@ contract StablecoinConverter is EpochTokenLocker {
+        // where maxUtility is the utility achieved when execSellAmount == order.sellAmount. That is,
+        // maxUtility = ((execBuyAmt * order.sellAmt - order.sellAmt * order.buyAmt) * price.buyToken) / order.sellAmt
+        //            = (execBuyAmt - order.buyAmount) * price.buyToken <---- This reduction not used in final formula.
+        // So, after simplification
```
Does it make sense to show the simplification here? It's not super obvious why execBuyAmt is no input to this function, but I guess the reader can also do the math himself?

bh2smith replied:
This was just included for my reviewers. Will disappear before merge!

fleupold commented on contracts/StablecoinConverter.sol:
```diff
@@ -209,20 +215,46 @@ contract StablecoinConverter is EpochTokenLocker {
+        //            = (execBuyAmt - order.buyAmount) * price.buyToken <---- This reduction not used in final formula.
+        // So, after simplification
+        (uint128 buyAmount, uint128 sellAmount) = getBuySellAmounts(order);
+        return (((sellAmount - execSell) * currentPrices[order.buyToken]) * buyAmount) / sellAmount;
```
Does this take into account if the account's balance would have been able to afford this trade? I think this is a crucial non-trivial requirement for that metric.

bh2smith replied:
I believe it does... since `submitSolution` would fail if it didn't.

josojo replied:
I believe it does not. All checks are done with regards to the volume, submitted as part of the solution if I recall correctly. But your function "just" looks at the order.remainingAmount

bh2smith replied:
and `order.remainingAmount` should be up to date with the balance. Unless it is not updated whenever deposits and withdraws are made/requested.

bh2smith replied:
If this is not the case, I would argue that this is an unrelated issue to the content of the PR.

fleupold reviewed (commented):
Cool stuff. Some more unit tests would be great to document how this metric behaves in certain cases (e.g. when disregarded orders are not touched or if the account doesn't have enough balance).

bh2smith committed:
Update formulas in smart contract and a couple other suggestions (not passing yet)

bh2smith committed:
revert disregarded utility evaluation chages, because it wasn't working.
Signed-off-by: Ben Smith <bh2smith@gmail.com>

bh2smith committed:
lint

bh2smith committed:
remove Debug events

bh2smith committed:
make relevant functions view and pure

bh2smith committed:
Merge branch 'master' into disregarded_utility

josojo commented on contracts/StablecoinConverter.sol:
```diff
@@ -209,20 +215,43 @@ contract StablecoinConverter is EpochTokenLocker {
             );
         }
-        checkAndOverrideObjectiveValue(uint(tokenConservation[0]));
+        int objectiveValue = utility - disregardedUtility;
```
Intuitiviely, I would go for uint instead of int for the objective value.

Because:
1. I can always submit the trivial solution without orders, which will have an objective value of 0
2. Any valuable solution in the 2-dimensional case has a positive utility- disregarded utility.
3. overflow issue handling with ints is much harder.

But for sure, we should check it more in detail

bh2smith replied:
Right, yes the trivial solution is considered "better" than other non-trivial ones with a lot of disregarded utility. This, kind-of, implies nobody would submit such a solution... because someone else would immediately submit the trivial solution which is better. In this case I would impose the following requirement

```js
require(utility >= disregardedUtility, "solution must be better than trivial");
```

bh2smith replied:
This is unfortunately not going to work as expected, since some of our unit tests have fairly large negative objective values.

josojo commented on contracts/StablecoinConverter.sol:
```diff
@@ -182,10 +183,15 @@ contract StablecoinConverter is EpochTokenLocker {
             require(checkOrderValidity(order, batchIndex), "Order is invalid");
             (uint128 executedBuyAmount, uint128 executedSellAmount) = getTradedAmounts(volumes[i], order);
+            // accumulate objective values before updateRemainingOrder
+            utility += evaluateUtility(executedBuyAmount, executedSellAmount, order);
```
please use the safe math operations

bh2smith replied:
yes good idea! Will do.

josojo commented on contracts/StablecoinConverter.sol:
```diff
@@ -209,20 +215,43 @@ contract StablecoinConverter is EpochTokenLocker {
             return previousSolution.objectiveValue;
         } else {
-            return 0;
+            return MIN_INT;
```
because of the previous comment, I guess we can keep 0.

But I am not sure

josojo commented on contracts/StablecoinConverter.sol:
```diff
@@ -209,20 +215,43 @@ contract StablecoinConverter is EpochTokenLocker {
+    function evaluateUtility(uint128 execBuy, uint128 execSell, Order memory order) internal view returns(int) {
+        // Utility = ((execBuyAmt * order.sellAmt - execSellAmt * order.buyAmt) * price.buyToken) / order.sellAmt
+        (uint128 buyAmount, uint128 sellAmount) = getBuyAndSellAmounts(order);
+        return (execBuy - ((execSell * buyAmount) / sellAmount)) * currentPrices[order.buyToken];
```
Why do you favor this solution over this one_

        return ((execBuy * sellAmount - execSell * buyAmount) / sellAmount)) * currentPrices[order.buyToken];

I don't know which one is better... Just asking stupid questions here :)

And please use safe math

bh2smith replied:
This alternative solution does not work because of the following example:
suppose `(a, b, c, d) = (1, 2, 3, 3)`
```
(a * b * c) / d = 6 / 3 = 2
BUT
((a * b) / d) * c = 0 * 3 = 0
```

Essentially, dividing first results in an undesired loss of precision. I have been working secretly on a Rational Numbers library in Solidity that might help with this stuff, but its not looking promising. Will bring it up tomorrow in the Smart Contracts BiWeekly

josojo commented on test/stablecoin_converter.js:
```diff
@@ -190,6 +190,7 @@ contract("StablecoinConverter", async (accounts) => {
       assert.equal((await stablecoinConverter.getBalance.call(user_1, erc20_2.address)).toNumber(), volume[0], "Bought tokens were not adjusted correctly")
       assert.equal((await stablecoinConverter.getBalance.call(user_2, erc20_2.address)).toNumber(), basicTrade.deposits[1].amount - getSellVolume(volume[1], prices[1], prices[0]), "Sold tokens were not adjusted correctly")
       assert.equal((await stablecoinConverter.getBalance.call(user_2, feeToken.address)).toNumber(), volume[1], "Bought tokens were not adjusted correctly")
+      // None of these assertions implies utility == 0?
```
Good observation. 

Now, you could check whether it is actually true.

josojo commented on test/stablecoin_converter.js:
```diff
@@ -1060,21 +1056,41 @@ contract("StablecoinConverter", async (accounts) => {
+      ]
+      const totalDisregardedUtility = disregardedUtilites.reduce((a, b) => a + b, 0)
+      const actualObjectiveValue = (await stablecoinConverter.getCurrentObjectiveValue.call()).toNumber()
+      assert.equal(actualObjectiveValue, totalUtility - totalDisregardedUtility, "Objective value is not stored correct")
```
really minor: usually, the expected value is the first value

bh2smith replied:
I believe exactly the opposite is true.

![Screenshot 2019-10-22 at 10 20 27 AM](https://user-images.githubusercontent.com/11778116/67268799-ad951f80-f4b5-11e9-82b9-82a85d6957cf.png)

![Screenshot 2019-10-22 at 10 20 49 AM](https://user-images.githubusercontent.com/11778116/67268794-ab32c580-f4b5-11e9-8837-5b970821f0c3.png)

josojo commented on test/stablecoin_converter.js:
```diff
@@ -1060,21 +1056,41 @@ contract("StablecoinConverter", async (accounts) => {
-      assert.equal((await stablecoinConverter.getCurrentObjectiveValue.call()).toNumber(), (getSellVolume(volume[0], prices[0], prices[1]) - volume[1]), "Objective value is not stored correct")
+      const tradeUtilities = [
+        evaluateTradeUtility(basicTrade.orders[0].buyAmount, basicTrade.orders[0].sellAmount, 4000, 8000, prices[1]),
+        evaluateTradeUtility(basicTrade.orders[1].buyAmount, basicTrade.orders[1].sellAmount, 7992, 4000, prices[0])
```
Could you maybe derive the values for the executed buyAmount and executedSellamout directly from the volume variables? I don't understand, why for the order[0], there is no fee subtraction happing, but for the order[1], there is.

josojo commented:
scraped

josojo commented:
This PR means great progress. I love it. You are pushing it @bh2smith!

bh2smith commented:
@josojo 
> dU_i = (((sellAmount - execSellAmount) * priceSellToken)

I like this idea because it would be a much simpler calculation. It probably wouldn't hurt to provide some justification for the equation or at least some way to move between our definition here and this one.

In other words, can we verify that
```py
(priceBuyToken * buyAmount) // sellAmount == priceSellToken
```

bh2smith committed:
Functional unit testing with only Utility (i.e. commented disregarded utility)

bh2smith committed:
Merge branch 'disregarded_utility' of github.com:gnosis/dex-contracts into disregarded_utility

bh2smith committed:
Fully functional U - dU with potentially missing unit tests.

bh2smith committed:
remove undefiend event emmision

bh2smith committed:
Update contract to use utility as objective function

bh2smith committed:
- Can't submit a solution with zero utility.
- Bunch of changes (in preparation for disregatded utility calculation)

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

# Conflicts:
#	contracts/StablecoinConverter.sol
#	test/stablecoin_converter.js

bh2smith committed:
no need for min int.

bh2smith committed:
Merge branch 'master' into utility_objective

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

bh2smith committed:
Use SafeMath and u256 in appropriate places

bh2smith committed:
small typo

bh2smith committed:
function can be restricted to view

bh2smith committed:
update revert message.

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

bh2smith committed:
use u256 before return

bh2smith committed:
No reason to have prices at 10.

bh2smith committed:
fix one test.

bh2smith committed:
fix one test.

bh2smith committed:
must verify limit price before evaluateUtility!

bh2smith committed:
must verify limit price before evaluateUtility!

josojo commented on contracts/StablecoinConverter.sol:
```diff
@@ -156,7 +156,7 @@ contract StablecoinConverter is EpochTokenLocker {
         address solutionSubmitter;
         uint256 feeReward;
-        uint256 objectiveValue;
+        uint objectiveValue;
```
I liked uint256 better :)

fleupold replied:
Shouldn't this be an int? Are we sure disregarded utility can never be larger than utility?

bh2smith replied:
I started with that, but the argument (which must have been offline) went as follows, the unsafty introduced with int is not worth the nuisance and since the trivial solution evaluates to zero, there is no point in having anything which evaluates below. 

Also notice the require statement below which enforces `utility > disregardedUtility`.

Until now, I have only managed to construct truly crappy, trivial,  examples whose objective is negative. For example, using the basicTrade consisting of two orders from the unit tests. One must fulfill 70% of each order\s requested buy amount in to meet the positivity threshold from our conditions. This suggests that fulfilment of these orders below 70% is considered a "worse solution" by our metric. 

Of course this is not exactly the ideal scenario for our objective function, but it does appear to capture quite a bit.

fleupold replied:
I'm convinced it's ok to be a unsigned integer, but I also agree that it should make the width explicit and thus stay an uint**256**

bh2smith replied:
I agree that `int` is more natural to consider in this scenario, but the argument about the trivial solution is pretty strong. Glad you are happy to use `uint`.

josojo reviewed (approved):
Really great!

Thanks for splitting the PRs. Much easier to review.

I think it is also great, that you do the normalization of the prices in the other PR.

fleupold commented on contracts/StablecoinConverter.sol:
```diff
@@ -244,6 +251,22 @@ contract StablecoinConverter is EpochTokenLocker {
     }
 
+    function evaluateDisregardedUtility(Order memory order, address user) internal view returns(uint128) {
+        // |disregardedUtility| = maxUtility - actualUtility
```
If this is is the definition, shouldn't we just call `evaluateUtility(remainingBuyAmount, order) - evaluateUtility(executedBuyAmount, order)`?

bh2smith replied:
Not quite, since utility involves a slightly different version of executedSellAmount and this is also no longer the formula being used. Thanks for pointing this out, I will update it!

bh2smith committed:
Merge branch 'master' into utility_objective

bh2smith committed:
Introduce scaled utility

bh2smith committed:
Merge branch 'utility_objective' of github.com:gnosis/dex-contracts into utility_objective

bh2smith committed:
update utility evaluation in JS test file

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

bh2smith committed:
remove commented line

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

bh2smith committed:
update comments in disregarded utility calculation.

bh2smith committed:
Max touched orders (#232)

In order to prevent an unfair advantage to first solution submitter, we must restrict the number of touched orders to something that only uses no more than 4 Million gas (i.e. approx 25 orders). This is based on the gas estimations made in #231

fleupold commented on contracts/StablecoinConverter.sol:
```diff
@@ -249,6 +255,22 @@ contract StablecoinConverter is EpochTokenLocker {
+        // TODO - use SafeCast
+        uint256 limitTerm = currentPrices[order.sellToken].mul(order.priceDenominator)
+            .sub(currentPrices[order.buyToken].mul(order.priceNumerator));
+        return uint128(leftoverSellAmount.mul(limitTerm)) / order.priceDenominator;
```
nit: SafeMath and shouldn't the cast happen after the division?

fleupold commented on contracts/StablecoinConverter.sol:
```diff
@@ -249,6 +255,22 @@ contract StablecoinConverter is EpochTokenLocker {
+        // where limitTerm = price.SellToken * order.sellAmt - order.buyAmt * price.buyToken
+        // and leftoverSellAmount = order.sellAmt - execSellAmt
+        // Balances and orders have all been updated so: sellAmount - execSellAmt == order.remainingAmount.
+        // For security reasons, we take the minimum of this with the user's token balance.
```
It's not really a security measure but just correctness, right?

bh2smith replied:
Yes, I suppose so!

bh2smith committed:
Update comment on utility evaluation and implement suggestion to pass fee rather than scale.

bh2smith committed:
include todo

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

# Conflicts:
#	contracts/StablecoinConverter.sol

bh2smith committed:
Merge branch 'disregarded_utility' of github.com:gnosis/dex-contracts into disregarded_utility

bh2smith committed:
update comment

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

bh2smith committed:
SafeMath and cast after division!

bh2smith committed:
update misleading comment

bh2smith committed:
easier objective evaluation

bh2smith committed:
Merge branch 'master' into utility_objective

fleupold commented on test/stablecoin_converter.js:
```diff
@@ -1137,7 +1148,40 @@ contract("StablecoinConverter", async (accounts) => {
         "sellAmount too large"
       )
     })
+    it("reverts if max touched orders is exceeded", async () => {
```
This should have probably been a separate PR?

bh2smith replied:
Do we not expect max touched orders to already have been included before this is merged?

bh2smith replied:
Ya... #232  have already been merged... oh ya... weird... I suppose this should already be in the merged PR...

fleupold commented on contracts/StablecoinConverter.sol:
```diff
@@ -249,6 +255,22 @@ contract StablecoinConverter is EpochTokenLocker {
 
+    function evaluateDisregardedUtility(Order memory order, address user) internal view returns(uint128) {
+        // |disregardedUtility| = (limitTerm * leftoverSellAmount) / order.sellAmount
+        // where limitTerm = price.SellToken * order.sellAmt - order.buyAmt * price.buyToken
```
Could you maybe explain what this "limitTerm" is supposed to represent?

bh2smith replied:
Yes certainly! But only after I fix the failing utility evaluation test!

fleupold reviewed (approved):
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia1.giphy.com%2Fmedia%2FiXQ8SgaMQAgtq%2Fgiphy.gif&token=lgtm" data-gitmeme-token="lgtm"><img src="https://media1.giphy.com/media/iXQ8SgaMQAgtq/giphy.gif" title="Created by gitme.me with /lgtm"/></a>

bh2smith committed:
Max touched orders (#232)

In order to prevent an unfair advantage to first solution submitter, we must restrict the number of touched orders to something that only uses no more than 4 Million gas (i.e. approx 25 orders). This is based on the gas estimations made in #231

# Conflicts:
#	test/stablecoin_converter.js

bh2smith committed:
Merge branch 'master' into utility_objective

# Conflicts:
#	test/stablecoin_converter.js

bh2smith committed:
update unit tests for utility funciton (no need to capture the fees with volume vector).

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

# Conflicts:
#	contracts/StablecoinConverter.sol
#	test/stablecoin_converter.js

bh2smith committed:
comment including rationality for burntFees in objective function!

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

# Conflicts:
#	contracts/StablecoinConverter.sol

bh2smith committed:
Update contracts/StablecoinConverter.sol

bh2smith committed:
Update contracts/StablecoinConverter.sol

bh2smith committed:
shorten comment for function num lines restriction.

bh2smith committed:
Update test/stablecoin_converter.js

bh2smith committed:
Update test/stablecoin_converter.js

bh2smith committed:
Merge branch 'master' into max_touched_orders

bh2smith committed:
Merge branch 'max_touched_orders' into utility_objective

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

# Conflicts:
#	test/stablecoin_converter.js

bh2smith committed:
Merge branch 'master' into utility_objective

# Conflicts:
#	test/stablecoin_converter.js

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

bh2smith committed:
Merge branch 'master' into disregarded_utility

bh2smith committed:
last fix
